File: src/IsbInterop.CSharp/ConfigurationUtils.cs
using System;
using System.ComponentModel;
using System.Configuration;
using System.IO;

namespace IsbInterop
{
  /// <summary>
  /// Утилиты для работы с конфигурацией.
  /// </summary>
  public static class ConfigurationUtils
  {
    /// <summary>
    /// Конфигурация isbInterop.
    /// </summary>
    private static readonly Lazy<Configuration> isbInteropConfiguration = new Lazy<Configuration>(TryGetIsbInteropConfiguration, true);

    /// <summary>
    /// Получить настройки строки подключения.
    /// </summary>
    /// <returns>Настройки строки подключения.</returns>
    public static string GetConnectionString()
    {
      ConnectionStringSettings connectionStringSettings;

      var standaloneConfiguration = isbInteropConfiguration.Value;
      if (standaloneConfiguration != null)
        connectionStringSettings = standaloneConfiguration.ConnectionStrings.ConnectionStrings["IsbInterop"];
      else
        connectionStringSettings = ConfigurationManager.ConnectionStrings["IsbInterop"];

      var connectionString = connectionStringSettings != null ?
        connectionStringSettings.ConnectionString : null;

      return connectionString;
    }

    /// <summary>
    /// Получить настройку секции appSettings.
    /// </summary>
    /// <typeparam name="T">Тип результата.</typeparam>
    /// <param name="settingName">Имя настройки/</param>
    /// <param name="result">Результат.</param>
    /// <returns>True, если удалось получить настройку, иначе false.</returns>
    public static bool TryGetAppSetting<T>(string settingName, ref T result)
    {
      if (settingName == null)
        throw new ArgumentNullException("settingName");

      string settingValue = null;

      var standaloneConfiguration = isbInteropConfiguration.Value;
      if (standaloneConfiguration != null)
      {
        var setting = standaloneConfiguration.AppSettings.Settings[settingName];
        if (setting != null)
          settingValue = setting.Value;
      }
      else
        settingValue = ConfigurationManager.AppSettings[settingName];

      if (settingValue != null)
        return TryConvert(settingValue, ref result);
      else
        return false;
    }

    /// <summary>
    /// Получить конфигурацию сборки IsbInterop.
    /// </summary>
    /// <returns>Конфигурация, если она существует, иначе null.</returns>
    private static Configuration TryGetIsbInteropConfiguration()
    {
      var currentAssembly = typeof(IsbApplicationManager).Assembly;
      var assemblyPath = new Uri(currentAssembly.CodeBase).LocalPath;
      var assemblyPathWithoutExtenstion = assemblyPath.Substring(0, assemblyPath.Length - 4);
      var configPath = string.Format("{0}.dll.config", assemblyPathWithoutExtenstion);

      // Проверяем, что файл существует, т.к.метод OpenMappedExeConfiguration всегда создает объект конфигурации.
      if (!File.Exists(configPath))
        return null;

      var configMap = new ExeConfigurationFileMap();
      configMap.ExeConfigFilename = configPath;

      Configuration configuration = null;
      try
      {
        configuration = ConfigurationManager.OpenMappedExeConfiguration(configMap, ConfigurationUserLevel.None);
      }
      catch (ConfigurationErrorsException)
      {
        // We could log.trace this exception if we have logger.
      }

      return configuration;
    }

    /// <summary>
    /// Сконвертировать строку в заданный тип.
    /// </summary>
    /// <typeparam name="T">Тип результата.</typeparam>
    /// <param name="source">Исходное значение.</param>
    /// <param name="result">Результат.</param>
    /// <returns>Экземпляр типа.</returns>
    private static bool TryConvert<T>(string source, ref T result)
    {
      var typeConverter = TypeDescriptor.GetConverter(typeof(T));
      if (typeConverter.CanConvertFrom(typeof(string)))
      {
        try
        {
          result = (T)typeConverter.ConvertFrom(source);
          return true;
        }
        // ReSharper disable once EmptyGeneralCatchClause
        catch (Exception)
        {
        }
      }

      return false;
    }

    /// <summary>
    /// Зашифровать строки подключения.
    /// </summary>
    public static void EncryptConnectionStrings()
    {
      var configuration = isbInteropConfiguration.Value;
      if (configuration == null)
      {
        try
        {
          configuration = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None);
        }
        catch (ConfigurationErrorsException)
        {
          return;
        }
      }

      var configSection = configuration.ConnectionStrings;

      if (configSection == null)
        return;

      try
      {
        if (!configSection.IsReadOnly() && !configSection.ElementInformation.IsLocked && !configSection.SectionInformation.IsLocked)
        {
          if (!configSection.SectionInformation.IsProtected)
            configSection.SectionInformation.ProtectSection("DataProtectionConfigurationProvider");

          configSection.SectionInformation.ForceSave = true;

          configuration.Save();
        }
      }
      // ReSharper disable once EmptyGeneralCatchClause
      catch (Exception)
      {
      }
    }
  }
}

File: src/IsbInterop/Utils/ConfigurationUtils.vb
Imports System.ComponentModel
Imports System.Configuration
Imports System.IO

Namespace Utils

  ''' <summary>
  ''' Утилиты для работы с конигурацией.
  ''' </summary>
  Public NotInheritable Class ConfigurationUtils

    ''' <summary>
    ''' Конфигурация isbInterop.
    ''' </summary>
    Private Shared ReadOnly isbInteropConfiguration As New Lazy(Of Configuration)(AddressOf TryGetIsbInteropConfiguration, True)

    ''' <summary>
    ''' Получат настройки строки подключения.
    ''' </summary>
    ''' <returns>Настройки строки подключения.</returns>
    Public Shared Function GetConnectionString() As String
      Dim connectionStringSettings As ConnectionStringSettings

      Dim standaloneConfiguration = isbInteropConfiguration.Value
      If standaloneConfiguration IsNot Nothing Then
        connectionStringSettings = standaloneConfiguration.ConnectionStrings.ConnectionStrings("IsbInterop")
      Else
        connectionStringSettings = ConfigurationManager.ConnectionStrings("IsbInterop")
      End If

      Dim connectionString = If(connectionStringSettings IsNot Nothing, connectionStringSettings.ConnectionString, Nothing)

      Return connectionString
    End Function

    ''' <summary>
    ''' Получает настройку секции appSettings.
    ''' </summary>
    ''' <typeparam name="T">Тип результата.</typeparam>
    ''' <param name="settingName">Имя настройки/</param>
    ''' <param name="result">Результат.</param>
    ''' <returns>True, если удалось получить настройку, иначе false.</returns>
    Public Shared Function TryGetAppSetting(Of T)(settingName As String, ByRef result As T) As Boolean
      If settingName Is Nothing Then
        Throw New ArgumentNullException("settingName")
      End If

      Dim settingValue As String = Nothing

      Dim standaloneConfiguration = isbInteropConfiguration.Value
      If standaloneConfiguration IsNot Nothing Then
        Dim setting = standaloneConfiguration.AppSettings.Settings(settingName)
        If setting IsNot Nothing Then
          settingValue = setting.Value
        End If
      Else
        settingValue = ConfigurationManager.AppSettings(settingName)
      End If

      If settingValue IsNot Nothing Then
        Return TryConvert(settingValue, result)
      Else
        Return False
      End If
    End Function

    ''' <summary>
    ''' Получает конфигурацию сборки IsbInterop.
    ''' </summary>
    ''' <returns>Конфигурация, если она существует, иначе null.</returns>
    Private Shared Function TryGetIsbInteropConfiguration() As Configuration
      Dim currentAssembly = GetType(IsbApplicationManager).Assembly
      Dim assemblyPath = New Uri(currentAssembly.CodeBase).LocalPath
      Dim assemblyPathWithoutExtenstion = assemblyPath.Substring(0, assemblyPath.Length - 4)
      Dim configPath = String.Format("{0}.dll.config", assemblyPathWithoutExtenstion)

      ' Проверяем, что файл существует, т.к.метод OpenMappedExeConfiguration всегда создает объект конфигурации.
      If Not File.Exists(configPath) Then
        Return Nothing
      End If

      Dim configMap = New ExeConfigurationFileMap()
      configMap.ExeConfigFilename = configPath

      Dim configuration As Configuration = Nothing
      Try
        configuration = ConfigurationManager.OpenMappedExeConfiguration(configMap, ConfigurationUserLevel.None)
        ' We could log.trace this exception if we have logger.
      Catch generatedExceptionName As ConfigurationErrorsException
      End Try

      Return configuration
    End Function

    ''' <summary>
    ''' Конвертирует строку в заданный тип.
    ''' </summary>
    ''' <typeparam name="T">Тип результата.</typeparam>
    ''' <param name="source">Исходное значение.</param>
    ''' <param name="result">Результат.</param>
    ''' <returns>Экземпляр типа.</returns>
    Private Shared Function TryConvert(Of T)(source As String, ByRef result As T) As Boolean
      Dim typeConverter = TypeDescriptor.GetConverter(GetType(T))
      If typeConverter.CanConvertFrom(GetType(String)) Then
        Try
          result = DirectCast(typeConverter.ConvertFrom(source), T)
          Return True
          ' ReSharper disable once EmptyGeneralCatchClause
        Catch ex As Exception
        End Try
      End If

      Return False
    End Function

    ''' <summary>
    ''' Шифрует строки подключения.
    ''' </summary>
    Public Shared Sub EncryptConnectionStrings()
      Dim configuration = isbInteropConfiguration.Value
      If configuration Is Nothing Then
        Try
          configuration = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None)
        Catch generatedExceptionName As ConfigurationErrorsException
          Return
        End Try
      End If

      Dim configSection = configuration.ConnectionStrings

      If configSection Is Nothing Then
        Return
      End If

      Try
        If Not configSection.IsReadOnly() AndAlso Not configSection.ElementInformation.IsLocked AndAlso Not configSection.SectionInformation.IsLocked Then
          If Not configSection.SectionInformation.IsProtected Then
            configSection.SectionInformation.ProtectSection("DataProtectionConfigurationProvider")
          End If

          configSection.SectionInformation.ForceSave = True

          configuration.Save()
        End If
        ' ReSharper disable once EmptyGeneralCatchClause
      Catch generatedExceptionName As Exception
      End Try
    End Sub

    ''' <summary>
    ''' Приватный конструктор.
    ''' </summary>
    Private Sub New()
    End Sub
  End Class
End NameSpace